enhancements/orange_okt_proposition.md

```diff
@@ -0,0 +1,326 @@
+
+1. I create a project with the Operator SDK command as usual
+2. In my Controller, I choose an OKT Reconciler and add the Reconciliation steps function in which I can see distinctly the standard steps in which I should pass at each reconciliation event (each step below is an entry in an embracing `switch(step) { case "xxx": ... }`:
+	- **CR Checker** - once the CR is picked up by OKT and validated by webhooks (if any), I can tell here if I have something to do more on my CR
+	- **ObjectGetter** - Here I will add the code to create my 3, in memory, resources and add them to the OKT's registry of resources (still in memory). However at this stage OKT pick up the resource on the K8S Cluster so it knows if the resource in its registry has an existing peer on the Cluster (or cache).
+	- **Mutator** - Here I tell to OKT to apply the Mutation on all resources present in its registry (load Initial/defaults data, then apply CR values if needed). 
+	- **Updater** - Here OKT, thanks to a hash algorithm on resources data, will compute if after Mutation, the resource has to be Created, Update, or is unchanged against it's cluster peer instance. So for all resource in its registry it apply the same idempotent process to update (or not!!) the resource.
+	- **SuccessManager** - If all goes right (no error) , this stage is reached, and here I'll can tell to OKT to "manageSuccess()" i.e. check for the right requeuing value to return and complete this reconciliation but at same time perform the update of a status condition for the Reconciliation, transparently for me, if I passed the CR.Status when I instatiate my OKT Reconciler object. 
+	- At any previous stage, if something goes wrong, an error can be raised, there's also the case where we giveup the reconciliation for whatever reason (not an error). There is also the case where a CR Finalization is triggered. These "debranching" cases are also present in my Reconciliation steps function thanks to 2 dedicated steps: 
+		-    **CRFinalizer** - here I can perform my own tasks when OKT detected that a finalizer (with the right name) exists and the CR is being deleted.
+		-    **ErrorManager** - This stage can be reached after any other stages, and let me call the OKT "manageError()" method that will pickup the last error and return the right requeuing value to complete this reconciliation.
+-    In the Controller folder I use the OKT code generator for my 3 resources I want to create. i.e.: 
+   `okt-gen-resource xxxxx`
+- In each file generated, I always find 3 parts to fill by myself and that will constitute a common/standard way to Mutate my resource:
+	- A method called `getHashXXX() { // Put your code here  }`  - It allows me to define, thanks to an OKT Helper, wich part of my object I want to include in the hash computation

```

Comment:
How is the hash's input created? Marhsaling the object itself? I've seen quite a few cases where a time or `resourceVersion` value is included in a hash/comparison, with an always-true result, so exposing such a function may be a source of confusion and bugs.

Reply:
Firstly for each resource, you have to define in the "getHashXXX()" method, the scope of objects/elements covered by the hash computation. OKT provides you an helper to define this scope. Internally it use an array of pointers with all the elements participating to the scope (the metadata you want, the whole Object's Spec part if you want, actually whatever you want). 
You return this "bag" of collected elements through the return statement of "getHashXXX()' method.

To perform the Hash computation, OKT use the spew library (`github.com/davecgh/go-spew/spew`) which follows pointers and prints actual values of the nested objects ensuring the hash does not change when a pointer changes. 
The returned hash can be used for object comparisons.

So if the developper don't add `resourceVersion` in the scope, there's no risk to meet the bug you mentionned. I know that some immutable fields (IP in Services, ...) and the Secret's data are things to take care about (OKT helper help on this point).
There is an example in the operator `Memcached with OKT` I sent to you and Tony (it can be shared).

Reply:
So it sounds like the hash is defined by the developer and they can choose as much or as little in their hash. And I agree with Eric that if someone unknowingly adds an attribute that changes it will cause an infinite reconcilation loop. (we've done this ourselves :)

Reply:
> if someone unknowingly adds an attribute that change it will cause an infinite reconcilation loop
OKT comes with an integrated throttling mechanism in case of a same recurrent errors, but not sure it can help in such case...
Besides that, when the developer defines the hash in the "getHashXXX()" method, he uses an OKT helper generic or dedicated for the resource kind he is dealing with.  For example, the SecretMutationHelper object manages few things with StringData and Data fields. It ensures too that the hash computation covers only the Data field and not the StringData field.
Here again, I don't know if there's a more elegant solutions and if a such helper is sufficient to guarantee the problem of the changing attributes added unknowingly.
The counterpart of these helpers is the need to maintain them. In OKT their usage is optional and today, the only one that exists, is the SecretMutationHelper. I thought to need another one for the Service resource and its changing ClusterIP, but in practice all works correctly without special care, it seems it was a mistake to think that.

Reply:
Edited the last sentence regarding specifics vs generics helpers.